### PR TITLE
fopen(): fix closing and errno

### PR DESCRIPTION
**Pull request by niewim19**
JIRA: PD-235

## Description

- Adding necessary file closing if `fopen()` fails to allocate memory after `safe_open()`
- move `mode == NULL` check to `string2mode()`
- accompany `string2mode()` change with correct errno returned when used

## Motivation and Context

- https://github.com/phoenix-rtos/phoenix-rtos-project/issues/298

## Types of changes

- [x] Bug fix (non-breaking change which fixes an issue)
- [ ] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to change)

## How Has This Been Tested?

- [ ] Already covered by automatic testing.
- [ ] New test added: (add PR link here).
- [x] Tested by hand on: ia32-generic.

## Checklist:

- [ ] My change requires a change to the documentation.
- [ ] I have updated the documentation accordingly.
- [ ] I have added tests to cover my changes.
- [x] All new and existing linter checks and tests passed.
- [x] My changes generate no new compilation warnings for any of the targets.

## Special treatment

- [ ] This PR needs additional PRs to work (list the PRs, preferably in merge-order).
- [ ] I will merge this PR by myself when appropriate.

### CODE REVIEW

File: stdio/file.c [hunk 1]
```diff
@@ -259,19 +256,21 @@ FILE *freopen(const char *pathname, const char *mode, FILE *stream)
 {
 	int m;
 
-	if (mode == NULL || stream == NULL)
+	if (stream == NULL) {
+		errno = EBADF;
```
Comment (agkaminski) [inline stdio/file.c:260]: _EBADF: The mode with which the file descriptor underlying the stream was opened does not support the requested mode when pathname is a null pointer._

This isn't really good return value in this case. Case of `stream` being `NULL` is not handled by the standard and most implementations don't check for `NULL` - they crash instead. This approach have it's benefits - user should check `stream` for `NULL` (after `fopen`), so `NULL` shouldn't be passed here ever. So this check is redundant and causes (very slight) performance hit.

IMHO - `EINVAL` or drop this check altogether.

Comment (niewim19) [reply stdio/file.c:260]: But how about:
>The freopen() function shall fail if:
[EBADF] The file descriptor underlying the stream is not a valid file descriptor when pathname is a null pointer. 

However I will change it to `EINVAL` as suggested. We need this check as then there is `fflush()` which will not throw an error when passed `NULL` an will flush `stdout`. So we might need this check to not flush stdout for no reason if function fails.

**Comment by niewim19** [comment]
Changes:
 - `freopen()` called with `stream = NULL` returns `EINVAL`

**Review by agkaminski — APPROVED** [review]
LGTM

**Comment by niewim19** [comment]
Dropped `freopen()` checking `stream == NULL` completely

**Pull request merged**